### PR TITLE
Add Windows support

### PR DESCRIPTION
**Pull request by kulibali**
This PR adds support for building on Windows, via a PowerShell script `make.ps1`.

- Updated the app and test code to handle Windows.  Note that on Windows, only `ponyup`, `ponyc`, and `corral` are supported packages.
- Instead of creating symbolic links, ponyup on Windows creates batch file shims in its bin directory (e.g. `ponyc.bat`) that call the selected versions with the given parameters.
- Added github actions for PR's, nightlies, and releases.
- Added a `ponyup-init.ps1` and updated the README to mention it.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -3,4 +3,4 @@
 /build/
 /cmd/version.pony
 /_corral/
-/_repos/
+/_repos/
```
Comment (SeanTAllen) [inline .gitignore:6]: i think this change can be reverted.

**Comment by ponylang-main** [comment]
Hi @kulibali,

The changelog - added label was added to this pull request; all PRs with a changelog label need to have release notes included as part of the PR. If you haven't added release notes already, please do.

Release notes are added by creating a uniquely named file in the `.release-notes` directory. We suggest you call the file `214.md` to match the number of this pull request.

The basic format of the release notes (using markdown) should be:

```
## Title

End user description of changes, why it's important,
problems it solves etc.

If a breaking change, make sure to include 1 or more
examples what code would look like prior to this change
and how to update it to work after this change.
```

Thanks.

File: cmd/ponyup.pony [hunk 2]
```diff
@@ -94,16 +94,18 @@ actor Ponyup
...
       return
     end
 
     (let install_path, let dl_path) =
       try
-        let p = _root.join(pkg'.string())?
-        (p, FilePath(_auth, p.path + ".tar.gz"))
+        let p = _root.join(pkg_str)?
+        (p, FilePath(_auth,
+          p.path + ifdef windows then ".zip" else ".tar.gz" end))
```
Comment (SeanTAllen) [inline cmd/ponyup.pony:108]: i think this would be easier to read if the string was constructed outside the `FilePath` call.

File: lock.json [hunk 3]
```diff
@@ -1,20 +1,32 @@
 {
   "locks": [
     {
-      "locator": "github.com/ponylang/crypto.git",
+      "locator": "github.com\\ponylang\\crypto.git",
```
Comment (SeanTAllen) [inline lock.json:4]: This looks to be incorrect.

File: lock.json [hunk 4]
```diff
@@ -1,20 +1,32 @@
...
       "revision": "1.1.3"
     },
     {
       "locator": "github.com/ponylang/appdirs.git",
       "revision": "0.1.3"
     },
     {
-      "locator": "github.com/ponylang/net_ssl.git",
-      "revision": "1.1.8"
+      "locator": "github.com\\ponylang\\http.git",
```
Comment (SeanTAllen) [inline lock.json:12]: this looks incorrect

File: lock.json [hunk 5]
```diff
@@ -1,20 +1,32 @@
...
     {
       "locator": "github.com/ponylang/http.git",
       "revision": "0.5.0"
+    },
+    {
+      "locator": "github.com/ponylang/crypto.git",
+      "revision": "1.1.3"
+    },
+    {
+      "locator": "github.com\\ponylang\\net_ssl.git",
```
Comment (SeanTAllen) [inline lock.json:24]: looks incorrect

File: lock.json [hunk 6]
```diff
@@ -1,20 +1,32 @@
...
+      "revision": "1.1.8"
+    },
+    {
+      "locator": "github.com\\ponylang\\appdirs.git",
```
Comment (SeanTAllen) [inline lock.json:28]: and here

File: test/main.pony [hunk 7]
```diff
@@ -88,20 +88,45 @@ class _TestSelect is UnitTest
     let install_args: {(String): Array[String] val} val =
       {(v) => ["update"; "ponyc"; v; "--platform=" + platform] }
 
-    let link =
-      FilePath(
-        h.env.root,
-        "./.pony_test/select/ponyup/bin/ponyc")
+    let link = FilePath(h.env.root, "./.pony_test/select/ponyup/bin/ponyc"
+      + ifdef windows then ".bat" else "" end)
```
Comment (SeanTAllen) [inline test/main.pony:92]: i think the string construction should be moved out of the FilePath call.

File: test/main.pony [hunk 8]
```diff
@@ -201,33 +232,52 @@ primitive _TestPonyup
   fun ponyup_bin(auth: AmbientAuth): FilePath? =>
     FilePath(auth, "./build")
       .join(if Platform.debug() then "debug" else "release" end)?
-      .join("ponyup")?
+      .join("ponyup" + ifdef windows then ".exe" else "" end)?
```
Comment (SeanTAllen) [inline test/main.pony:235]: This is hard to read.

File: test/main.pony [hunk 9]
```diff
@@ -201,33 +232,52 @@ primitive _TestPonyup
...
+      dbg_str
+    end)
+
     let ponyup_monitor = ProcessMonitor(
       auth,
       auth,
       object iso is ProcessNotify
         fun stdout(process: ProcessMonitor ref, data: Array[U8] iso) =>
-          h.log(String.from_array(consume data))
+          h.env.out.print(String.from_array(consume data))
```
Comment (SeanTAllen) [inline test/main.pony:261]: why the change?

File: README.md [hunk 10]
```diff
@@ -18,10 +18,18 @@ brew install libressl
...
 
+On Unix:
+
 ```bash
 sh -c "$(curl --proto '=https' --tlsv1.2 -sSf https://raw.githubusercontent.com/ponylang/ponyup/latest-release/ponyup-init.sh)"
 ```
 
+On Windows PowerShell:
+
+```
```
[the remark this answers is not in the document]
Comment (kulibali) [reply README.md:29]: `pwsh.exe` is the file name of the PowerShell 7 executable.

File: .gitignore (hunk 1 above)
Comment (kulibali) [reply .gitignore:6]: Why? I don't think this directory should be in git.

File: README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (SeanTAllen) [reply README.md:29]: That's not a valid code fence type. I believe it should be pwsh otherwise you'll fail MD linting.

File: .gitignore (hunk 1 above)
Comment (SeanTAllen) [reply .gitignore:6]: Sorry, what? You removed a new line. That's the only change here. 

Are you saying git ignore shouldn't be checked in? I would disagree with that.

File: cmd/main.pony [hunk 11]
```diff
@@ -20,7 +20,7 @@ actor Main is PonyupNotify
       return
     end
 
-    if not Platform.posix() then
+    if not (Platform.posix() or Platform.windows()) then
```
[the remark this answers is not in the document]
Comment (kulibali) [reply cmd/main.pony:23]: None at the moment. But this is future-proof in that it won't try to support future platforms.

File: README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (kulibali) [reply README.md]: Fixed.

File: README.md [hunk 12]
```diff
@@ -38,7 +46,9 @@ These commands will download the chosen version of ponyc and install it to `$HOM
 
 ### Set install prefix
 
-By default, ponyup will create its root directory in `$HOME/.local/share`. This prefix can be set manually with the `--prefix` (or `-p`) option. All packages selected as default will be symbolically linked into `${prefix}/ponyup/bin`. So, by default, `ponyup update release ponyc` will install `ponyc` to `$HOME/.local/share/ponyup/bin/ponyc`.
+By default, ponyup will create its root directory in `$HOME/.local/share` (`%LOCALAPPDATA%\ponyup` on Windows). This prefix can be set manually with the `--prefix` (or `-p`) option. All packages selected as default will be symbolically linked into `${prefix}/ponyup/bin`. So, by default, `ponyup update release ponyc` will install `ponyc` to `$HOME/.local/share/ponyup/bin/ponyc`.
+
+On Windows, ponyup will create batch file shims (e.g. `ponyc.bat` in `%LOCALAPPDATA%\ponyup\bin`) that will run the selected version.
```
[the remark this answers is not in the document]
Comment (kulibali) [reply README.md]: Done.

File: cmd/http_handlers.pony [hunk 13]
```diff
@@ -110,7 +110,11 @@ actor DLDump
       let progress_bar = recover String end
       progress_bar.append("\r  |")
       for i in Range(0, 100, 2) do
-        progress_bar.append(if i <= percent then "█" else "-" end)
+        progress_bar.append(if i <= percent then
+          ifdef windows then "#" else "█" end
+        else
+          "-"
+        end)
```
[the remark this answers is not in the document]
Comment (kulibali) [reply cmd/http_handlers.pony]: Done.

File: cmd/ponyup.pony [hunk 14]
```diff
@@ -94,16 +94,18 @@ actor Ponyup
       end
 
     let pkg' = (consume pkg).update_version(version)
+    let pkg_str: String val = pkg'.string()
 
     if _lockfile.contains(pkg') then
-      _notify.log(Info, pkg'.string() + " is up to date")
+      _notify.log(Info, pkg_str + " is up to date")
```
[the remark this answers is not in the document]
Comment (kulibali) [reply cmd/ponyup.pony:100]: It was for debugging.

File: cmd/ponyup.pony (hunk 2 above)
Comment (kulibali) [reply cmd/ponyup.pony]: Fixed.

File: lock.json (hunk 3 above)
Comment (kulibali) [reply lock.json]: Hmm, this would be a bug in corral. Fixed.

File: lock.json (hunk 4 above)
Comment (kulibali) [reply lock.json]: Fixed.

File: lock.json (hunk 5 above)
Comment (kulibali) [reply lock.json]: Fixed.

File: lock.json (hunk 6 above)
Comment (kulibali) [reply lock.json]: Fixed.

File: test/main.pony (hunk 7 above)
Comment (kulibali) [reply test/main.pony]: Done.

File: test/main.pony (hunk 8 above)
Comment (kulibali) [reply test/main.pony]: Fixed.

File: test/main.pony (hunk 9 above)
Comment (kulibali) [reply test/main.pony]: For debugging; I've reverted it.

File: cmd/main.pony (hunk 11 above)
[the remark this answers is not in the document]
Comment (SeanTAllen) [reply cmd/main.pony:23]: Are we future proofing everywhere else? I think it makes sense to either future proof or not. I don't believe we are future proofing elsewhere. If you think it is worth it, then I think reworking the code some to catch unsupported platforms easier would be a good ticket to open.

File: lock.json (hunk 3 above)
Comment (SeanTAllen) [reply lock.json]: Can you open an issue against corral?

**Comment by SeanTAllen** [comment]
I'm personally very excited about this PR.

File: .gitignore (hunk 1 above)
Comment (kulibali) [reply .gitignore:6]: Oh, I misread this change, I thought that the `_repos` directory had been removed from `.gitignore` file and I thought I added it back.  I'll revert the change.

File: cmd/ponyup.pony (hunk 14 above)
[the remark this answers is not in the document]
Comment (SeanTAllen) [reply cmd/ponyup.pony:100]: Are you going to revert or leave as is?

If leaving, please go ahead and resolve this conversation.

**Comment by SeanTAllen** [comment]
@kulibali let me know when you are ready for another review.

File: cmd/main.pony (hunk 11 above)
[the remark this answers is not in the document]
Comment (kulibali) [reply cmd/main.pony]: Updated.

**Comment by kulibali** [comment]
Hmm, on this one `ponyc` is exiting with code -1 but no error messages.

**Comment by SeanTAllen** [comment]
I dont see a failure from CI @kulibali, is there a problem hidden behind the green?

**Comment by kulibali** [comment]
I re-ran the test and it passed.  Not sure why Windows is being super flaky all of a sudden.

**Comment by SeanTAllen** [comment]
@kulibali did you see the change in windows version CI runner I sent in the CI channel on Zulip the other day? Perhaps that is it?

https://ponylang.zulipchat.com/#narrow/stream/190359-ci/topic/New.20Windows.20version.20for.20runners
